File: app/services/position/position_service.py
from uuid import uuid4

from app.models.position import Position

from app.repositories.position_repository import (
    PositionRepository
)


class PositionService:

    def __init__(
        self,
        db
    ):
        self.repo = PositionRepository(db)

    async def create_position(
        self,
        symbol: str,
        quantity: float,
        average_price: float,
        stop_loss: float,
        take_profit: float
    ):

        position = Position(
            id=str(uuid4()),
            symbol=symbol,
            quantity=quantity,
            average_price=average_price,
            stop_loss=stop_loss,
            take_profit=take_profit,
            unrealized_pnl=0
        )

        return await self.repo.create(
            position
        )

    async def get_position(
        self,
        symbol: str
    ):

        return await self.repo.get_by_symbol(
            symbol
        )

    async def get_all_positions(
        self
    ):

        return await self.repo.get_all()

    async def close_position(
        self,
        symbol: str
    ):

        position = await self.repo.get_by_symbol(
            symbol
        )

        if position:

            await self.repo.delete(
                position
            )

    async def update_position(
        self,
        position
    ):

        await self.repo.update()

        return position

File: backend/src/services/positionService.ts
import { randomUUID } from "crypto";
import { PositionRepository } from "../repositories/positionRepository";

export class PositionService {
  static async createPosition(
    symbol: string,
    quantity: number,
    averagePrice: number,
    stopLoss: number,
    takeProfit: number
  ) {
    return PositionRepository.create({
      id: randomUUID(),
      symbol,
      quantity,
      average_price: averagePrice,
      stop_loss: stopLoss,
      take_profit: takeProfit,
      unrealized_pnl: 0,
    });
  }

  static async getPosition(symbol: string) {
    return PositionRepository.getBySymbol(symbol);
  }

  static async getAllPositions() {
    return PositionRepository.getAll();
  }

  static async closePosition(symbol: string) {
    const position = await PositionRepository.getBySymbol(symbol);
    if (position) {
      await PositionRepository.delete(symbol);
    }
  }

  static async updatePosition(symbol: string, data: {
    quantity?: number;
    average_price?: number;
    stop_loss?: number | null;
    take_profit?: number | null;
    unrealized_pnl?: number;
  }) {
    return PositionRepository.update(symbol, data);
  }
}
export default PositionService;
